KAFKA-3490; Multiple version support for ducktape performance tests

ijuma opened the pull request.

Commit:
Initial sketch of versioned performance services

Commit:
Added version-awareness to end to end latency

Commit:
Added version support to consumer performance service

Commit:
ConsumerPerformance: override new consumer flag if version < 0.9.0.0

Commit:
Added sanity check on consumer performance with new consumer

Commit:
Updated ProducerPerformanceService - 0.8.X now runs 0.9 ProducerPerformance.java using 0.8.X java producer

Commit:
Added logging to ProducerPerformanceService

Commit:
Updated comment

Commit:
Updated producer performance to parse output from file after the service has finished running

Commit:
Added stderr capture to producerperformance services

Commit:
Updated producer perf to gather jmx stats

ijuma commented:
I revised @granders branch and rebased it against trunk. Running the branch builder here:

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/401/

@granders, if you could take a look that would be great.

ijuma commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -234,6 +234,13 @@ public void printTotal() {
+      6: mb/s, 7: totalMessagesSent.get, 8: msgs/sec))
+             * *
+             */
+            // 2015-11-17 03:46:42:901, 2015-11-17 03:46:43:158, 0, 100, 200, 0.00, 0.0037, 10, 38.9105\n',)}
```
I will remove this.

granders commented on tests/kafkatest/services/performance/end_to_end_latency.py:
```diff
@@ -36,17 +38,23 @@ def __init__(self, context, num_nodes, kafka, topic, num_records, consumer_fetch
+        args = self.args.copy()
         args.update({
             'zk_connect': self.kafka.zk.connect_setting(),
             'bootstrap_servers': self.kafka.bootstrap_servers(self.security_config.security_protocol),
```
Line 61:
I think we want `'security_config_file': self.security_config_file`? `security_config_file` looks like it's out of scope now

ijuma replied:
Fixed.

ijuma committed:
Remove commented out code in `ProducerPerformance`.

ijuma commented:
Started a system tests build here: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/405/

ijuma commented on tests/kafkatest/services/performance/consumer_performance.py:
```diff
@@ -127,15 +135,31 @@ def start_cmd(self, node):
-        cmd += " --consumer.config %s" % ConsumerPerformanceService.CONFIG_FILE
 
-        for key, value in self.settings.items():
-            cmd += " %s=%s" % (str(key), str(value))
```
@granders, is it right that these two lines should be removed?

granders replied:
@ijuma Good catch - I agree these should not be removed

ijuma commented on tests/kafkatest/services/performance/producer_performance.py:
```diff
@@ -13,21 +13,45 @@
+            "path": LOG_FILE,
             "collect_default": True},
+        "jmx_log": {
+            "path": "/mnt/jmx_tool.log",
```
@granders, is it intentional that the jmx log is not under `PERSISTENT_ROOT`?

granders replied:
@ijuma I don't think so - maybe an artifact of cherry-picking?

ijuma replied:
I think this is because `/mnt/jmx_tool.log` is hardcoded in `jmx.py`.

granders replied:
After looking into the WARNINGs you mentioned collecting logs, it might make sense to set the jmx_log collect_default to True iff `jmx_object_names is not None`

ijuma replied:
Thanks for looking into that, will do.

ijuma committed:
Add 0.9 version and change default in performance service sanity check

granders commented on tests/kafkatest/sanity_checks/test_performance_services.py:
```diff
@@ -0,0 +1,86 @@
+    def setUp(self):
+        self.zk.start()
+
+    @parametrize(version=str(LATEST_0_8_2))
```
If 0.9.0.X is the current baseline, maybe we don't need to run 0.8.2.X every time?

granders replied:
Maybe an additional comment here to justify why we would keep the 0.8.2.X test case around for the time being?

ijuma committed:
Add back code that updates `cmd` from `settings`

granders commented on tests/kafkatest/services/performance/consumer_performance.py:
```diff
@@ -78,6 +78,11 @@ def __init__(self, context, num_nodes, kafka, topic, messages, new_consumer=Fals
+        if version < V_0_9_0_0 and self.new_consumer:
+            self.logger.warn("Version is %s where new_consumer is not available, but new_consumer is set to True. "\
+                             "Overriding new_consumer flag to False..." % str(version))
+            self.new_consumer = False
```
Looking back at this, it's a bit fishy - maybe better to enforce this with an assertion rather than warning and modifying the `self.new_consumer`

granders commented on tests/kafkatest/services/performance/consumer_performance.py:
```diff
@@ -127,15 +135,31 @@ def start_cmd(self, node):
-            cmd += " %s=%s" % (str(key), str(value))
+        if node.version > LATEST_0_8_2:
+            # This is only used for security settings
+            cmd += " --consumer.config %s" % ConsumerPerformanceService.CONFIG_FILE
```
On second look - `cmd += " --consumer.config %s" % ConsumerPerformanceService.CONFIG_FILE` was simply moved here

ijuma replied:
Yeah, that's why I only asked about the two lines copying the `settings`. So, I was going to add it back after this `if` block. Is that correct?

granders commented on tests/kafkatest/services/performance/end_to_end_latency.py:
```diff
@@ -17,9 +17,11 @@
 
 
 class EndToEndLatencyService(PerformanceService):
+    MESSAGE_BYTES = 21  # 0.8.X messages are fixed at 21 bytes, so we'll match that for other versions
```
It's probably worth double-checking whether this assertion is true

From what I remember, the 0.8.X version of EndToEndLatency sends a hard-coded message, and I computed it at 21 bytes, but it's possible that past Geoff made a mistake

ijuma replied:
Past Geoff was correct.

granders commented on tests/kafkatest/services/performance/end_to_end_latency.py:
```diff
@@ -36,28 +38,65 @@ def __init__(self, context, num_nodes, kafka, topic, num_records, consumer_fetch
             'consumer_fetch_max_wait': consumer_fetch_max_wait,
             'acks': acks,
-            'kafka_opts': self.security_config.kafka_opts
+            'kafka_opts': self.security_config.kafka_opts,
```
It may be worth validating that security protocol is plaintext if version is less than 0.9.0.0
Same comment for other services where it isn't already enforced

ijuma committed:
Merge remote-tracking branch 'apache/trunk' into kafka-3490-multiple-version-support-perf-tests

* apache/trunk:
  KAFKA-3477: extended KStream/KTable API to specify custom partitioner for sinks
  KAFKA-3508: Fix transient SimpleACLAuthorizerTest failures
  KAFKA-3489; Update request metrics if a client closes a connection while the broker response is in flight
  KAFKA-3510; OffsetIndex thread safety

ijuma committed:
Explain why we still have `LATEST_0_8_2` in `test_performance_services.py`

ijuma committed:
Add asserts related to the new consumer and security protocols

As suggested by Geoff

ijuma committed:
Add version support to `benchmark_test.py`

ijuma committed:
Only collect_default for jmx_tool.log if jmx_object_names is not None

ijuma commented:
@granders, ready for another round of reviews. I addressed your feedback and added version support to `benchmark_test.py`. I don't have any other changes planned.

I tested the relevant tests via the AWS test driver and things seem fine. I also started a full system tests run via the branch builder:

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/407/

ijuma commented:
System tests passed.

@ewencp, it would also be good for you to take a look as I assume you'll be the committer reviewing this. Some context: Geoff did most of the work on this branch before 0.9.0.0 went out. I needed this functionality, so Geoff rebased the branch and I tested it and added minor fixes and improvements.

granders commented on tests/kafkatest/services/performance/producer_performance.py:
```diff
@@ -44,56 +74,103 @@ def __init__(self, context, num_nodes, kafka, topic, num_records, record_size, t
+
+        cmd = ""
+
+        if node.version <= LATEST_0_8_2:
```
@I think this should actually probably be:
`if node.version < TRUNK:`

granders replied:
amended: make that strictly less than

ijuma replied:
Fixed it locally. Testing it before I push (as we discussed, I noticed this issue as well when running a performance test with a change to `ProducerPerformance`).

granders commented:
@ijuma Other than fixing the conditional in `producer_performance.py` this looks good to me. Let's see if @ewencp has further comments

ijuma committed:
Fix version logic to always use `ProducerPerformance` from trunk

ewencp commented on tests/kafkatest/services/performance/producer_performance.py:
```diff
@@ -44,56 +74,103 @@ def __init__(self, context, num_nodes, kafka, topic, num_records, record_size, t
-        last = None
+        # start ProducerPerformance process
         producer_output = node.account.ssh_capture(cmd)
+        wait_until(lambda: self.alive(node), timeout_sec=20, err_msg="ProducerPerformance failed to start")
```
@ijuma This seems like it might not work. In practice, all tests that use this should have it run long enough for this to pass (or correctly fail). But your sanity checks seem like a case where this could fail -- the process could startup and manage to exit before the first check for liveness runs.

granders replied:
@ewencp Isn't this true of any check of "aliveness" here since the process can be gone by the time you check?

In principle you'd need to detect
"is alive OR was alive and finished successfully"

One option is to simply keep as-is, but document that there a known (but probably unlikely) way to get a spurious failure here

ewencp replied:
@granders Yes, what you say is true, but in practice this type of test _normally_ works if you expect the process to be running for more than a few seconds. Normally I wouldn't have called this out, but since the sanity check tests that were added specifically try to be short-lived, I want to make sure we're not likely to encounter transient failures on them.

granders replied:
Fair enough - my thought was that maybe it's ok to keep it simple (but document) for the time being since we aren't seeing these transient failures?

ewencp commented on tests/kafkatest/services/performance/templates/tools_log4j.properties:
```diff
@@ -19,7 +19,7 @@ log4j.rootLogger = {{ log_level|default("INFO") }}, FILE
-# Set the append to false, overwrite
-log4j.appender.FILE.Append=false
+# Set the append to true, append
+log4j.appender.FILE.Append=true
```
Why was this needed? Append is normal for long running services that we need to bounce since we want to capture all the output, but this seems unusual for tools. They are usually one-off. I don't think there's necessarily anything wrong with this, but I'd like to understand the reasoning since this template is used across 5 or 6 separate tools.

granders replied:
I may have made this change in the original branch back in December, and I honestly don't remember if I had a good reason for it.

I agree it's probably safer to revert this since several tools rely on it.

ijuma committed:
Replace `version` with `client_version` and `broker_version` in `benchmark_test.py`

ijuma committed:
Set file append back false in tools_log4j.properties

ijuma committed:
Merge remote-tracking branch 'apache/trunk' into kafka-3490-multiple-version-support-perf-tests

* apache/trunk:
  MINOR: improve logging of consumer system tests
  KAFKA-725: Return OffsetOutOfRange error from ReplicaManager when non-follower attempts reading an offset that's above high watermark.
  KAFKA-3512: Added foreach operator
  KAFKA-3505: Fix punctuate generated record metadata
  MINOR: ensure original use of prop_file in verifiable producer
  KAFKA-3488; Avoid failing of unsent requests in consumer where possible
  KAFKA-3497: Streams ProcessorContext should support forward() based on child name

ijuma commented:
@ewencp Running the sanity check test in a loop of 100 iterations here:

https://jenkins.confluent.io/job/system-test-kafka-branch-builder-2/49/consoleFull

So far, 20 iterations passed.

ijuma commented:
OK, so the test passed 67 times and then it failed with what I think is an unrelated failure:

[INFO:2016-04-09 03:35:02,127]: SerialTestRunner: Summary: There aren't enough available nodes to satisfy the resource request. Total cluster size: 0, Requested: 1, Already allocated: 0, Available: 0. Make sure your cluster has enough nodes to run your test or service(s). Currently registered services: [<ZookeeperService: num_nodes: 1, nodes: []>]

ewencp commented:
@ijuma That actually sounds like something went wrong with getting the list of machines available in the cluster (or they shutdown or something). This is what you normally get if you don't have any nodes started, or if you just don't have enough nodes for the test if the numbers are larger. 67 passes is enough to convince me :)

LGTM
